lib/virtual-fs/src/static_file.rs

```diff
@@ -30,49 +31,44 @@ impl VirtualFile for StaticFile {
     fn last_accessed(&self) -> u64 {
         0
     }
+
     fn last_modified(&self) -> u64 {
         0
     }
+
     fn created_time(&self) -> u64 {
         0
     }
+
     fn size(&self) -> u64 {
-        self.len
+        self.0.get_ref().len().try_into().unwrap()
     }
-    fn set_len(&mut self, _new_size: u64) -> crate::Result<()> {
-        Ok(())
+
+    fn set_len(&mut self, _new_size: u64) -> Result<(), FsError> {
+        Err(FsError::PermissionDenied)
     }
+
     fn unlink(&mut self) -> BoxFuture<'static, Result<(), FsError>> {
-        Box::pin(async { Ok(()) })
+        Box::pin(async { Err(FsError::PermissionDenied) })
     }
+
     fn poll_read_ready(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<io::Result<usize>> {
-        let remaining = self.len - self.cursor;
-        Poll::Ready(Ok(remaining as usize))
+        let remaining = self.0.position() - self.size();
+        Poll::Ready(Ok(remaining.try_into().unwrap()))
     }
+
     fn poll_write_ready(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<io::Result<usize>> {
-        Poll::Ready(Ok(0))
+        Poll::Ready(Err(std::io::ErrorKind::PermissionDenied.into()))
     }
 }
 
 impl AsyncRead for StaticFile {
     fn poll_read(
-        self: Pin<&mut Self>,
-        _cx: &mut Context<'_>,
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
         buf: &mut tokio::io::ReadBuf<'_>,
     ) -> Poll<io::Result<()>> {
-        let bytes = self.bytes.as_ref();
-
-        let cursor: usize = self.cursor.try_into().unwrap_or(u32::MAX as usize);
-        let _start = cursor.min(bytes.len());
-        let bytes = &bytes[cursor..];
-
-        if bytes.len() > buf.remaining() {
-            let remaining = buf.remaining();
-            buf.put_slice(&bytes[..remaining]);
-        } else {
-            buf.put_slice(bytes);
-        }
-        Poll::Ready(Ok(()))
+        Pin::new(&mut self.0).poll_read(cx, buf)
```

Comment:
Nit: this is usually done with `x.as_mut()` (method on Pin)

Also applies to the other uses.